examples/pissa_finetuning/pissa_finetuning.py

```diff
@@ -0,0 +1,172 @@
+        script_args.residual_model_name_or_path, quantization_config=quantization_config, low_cpu_mem_usage=True
+    )
+    res_model = prepare_model_for_kbit_training(res_model)
+    print("Wrapping the residual model with PiSSA.")
+    peft_model = PeftModel.from_pretrained(
+        res_model, script_args.residual_model_name_or_path, subfolder="pissa_init", is_trainable=True
+    )
+    tokenizer = AutoTokenizer.from_pretrained(script_args.residual_model_name_or_path)
+
+elif script_args.residual_model_name_or_path is not None:
+    res_model = AutoModelForCausalLM.from_pretrained(
+        script_args.residual_model_name_or_path,
+        torch_dtype=(
+            torch.float16
+            if script_args.bits == "fp16"
+            else (torch.bfloat16 if script_args.bits == "bf16" else torch.float32)
+        ),
+        device_map="auto",
+    )
+    print("Wrapping the residual model with PiSSA.")
+    peft_model = PeftModel.from_pretrained(
+        res_model, script_args.residual_model_name_or_path, subfolder="pissa_init", is_trainable=True
+    )
+    tokenizer = AutoTokenizer.from_pretrained(script_args.residual_model_name_or_path)
+
+elif script_args.base_model_name_or_path is not None:
+    print(
+        f"No available pre-processed model, manually initialize a PiSSA using {script_args.base_model_name_or_path}."
+    )
+    model = AutoModelForCausalLM.from_pretrained(
+        script_args.base_model_name_or_path,
+        torch_dtype=(
+            torch.float16
+            if script_args.bits == "fp16"
+            else (torch.bfloat16 if script_args.bits == "bf16" else torch.float32)
+        ),
+        device_map="auto",
+    )
+    tokenizer = AutoTokenizer.from_pretrained(script_args.base_model_name_or_path)
+    tokenizer.pad_token_id = tokenizer.eos_token_id
+    lora_config = LoraConfig(
+        r=script_args.lora_r,
+        lora_alpha=script_args.lora_alpha,
+        init_lora_weights=script_args.init_lora_weights,
+        lora_dropout=script_args.lora_dropout,
+        target_modules=["q_proj", "o_proj", "k_proj", "v_proj", "gate_proj", "up_proj", "down_proj"],
+        bias="none",
+        task_type="CAUSAL_LM",
+    )
+    peft_model = get_peft_model(model, lora_config)
+
+    if script_args.save_for_reuse:
+        init_model = deepcopy(peft_model)
+        # Save PiSSA modules:
+        init_model.peft_config["default"].init_lora_weights = True
+        init_model.save_pretrained(os.path.join(script_args.output_dir, "pissa_init"))
+        # Save residual model:
+        init_model = init_model.unload()
+        init_model.save_pretrained(script_args.output_dir)
+        # Save the tokenizer:
+        tokenizer.save_pretrained(script_args.output_dir)
+else:
+    raise ValueError("At least one of `base_model_name_or_path` and `residual_model_name_or_path` must not be None.")
+
+print(peft_model)
+peft_model.print_trainable_parameters()
+if script_args.do_train:
```

Comment:
If a user passes `do_train=False` and `save_for_reuse=False`, this script essentially does nothing, right? I wonder if we should check for this and give a helpful hint to the user. Alternatively, we could split this script into 2 scripts, one for running what is essentially lines 124-135, and one for training (and thus get rid of the two mentioned options).